### ISSUE TITLE
Metadata within STACK questions

### ISSUE THREAD

**Issue by sangwinc** [maths/moodle-qtype_stack#1171]
At the STACK International Advisory Board meeting today, one colleague asked for a supported mechanism for question authors to add metadata to STACK questions.  E.g. author(s), original institution, copyright, etc. This could be an extra field.

The first thing to note is this really isn't an issue only affecting STACK questions.  But, since STACK questions are more complicated than many, people see more value in these than other question types.

One proposal is to add in a [[metadata]] ... [[/metadata]] castext block which, like comments, is not shown to students.  Instead it can be used to host metadata.

Newer versions of STACK have a "description" field, and the block would naturally sit there.

Do people want to add this data?  What do you want to add, and how can we easily add it in a consistent and simple way?

**Comment by bgailer** [comment]
I think that is a wonderful feature that would be beneficial for the use case at our university. In our STACK questions we usually include information about the author, institution and also sometimes the copyright as a comment in the question variables, but this is surely not the best place for this kind of information. 

The question description field would be a good place in my opinion. Maybe every new question could start with an empty [[metadata]] block in the question description filed? Also, maybe the content of that [[metadata]] castext block could be shown below the question name in the question dashboard so that one can quickly see who created a question and who modified it  (especially useful for imported question). I'm not sure if a new field for question metadata would be better that a castext block.

Or maybe something like this:
```
[[metadata name="" role="author" institution="" copyright="" /]]
[[metadata name="" role="editor" institution="" copyright="" /]]
```

Overall, this proposal would definitely help for dealing with author information on questions :)

**Comment by smborio** [comment]
Hi all,

We already use the question description for this. We have the following tags that match our workflows (and are currently discussing the possibility of adding some more):

```
<p id="author"></p>
<p id="concept doc link"></p>
<p id="reviewer"></p>
<p id="reviewer"></p>
<p id="description"></p>
```

I hope this helps in the definition of this addition. 

There is another type of metadata that we are in the process of using, which is already included, which is tags. This is to help us manage large question banks and with gitsync we are hoping to have a sensible database that uses the metadata to filter questions nicely.

All the best!

**Comment by sangwinc** [comment]
One option would be to have a tag to indicate the _type_ of metadata.  If we support JSON then we could more easily compute with it later, e.g. 

```
[[metadata type="json"]] 
{
 "author": "Chris Sangwin",
 "licence": "CCBY"
}
[[/metadata]]

**Comment by sangwinc** [comment]
What metadata should we store?  My suggestion is that we add another setting to STACK which is the default metadata for a new question.  
````
qtype_stack | metadata
```` 
When we create a new question, this becomes the question description, and some of the fields in this metadata (e.g. question author) are automatically populated.

We might investigate existing standards for metadata as well, e.g https://pure.hw.ac.uk/ws/portalfiles/portal/10607125/TICLpaper.MetadataForEducation_postref.pdf
including The Dublin Core Metadata Initiative (DCMI).

If we really go for Dublin Core, then we might validate this with
````
[[metadata type="dublin"]] 
````

**Comment by sangwinc** [comment]
Yes, I agree we should have a separate metadata field in the STACK question.  Unless moodle plan to do this soon (I'm checking) then we should add this to the STACK question type immediately.  We now have many questions banks, material development projects, and legacy collections which really need some basic metadata immediately.

**Comment by aharjula** [comment]
@EJMFarrow excellent picture. Some more complicated structures for well-known bits of meta, and still the option of putting "Additional Metadata" for everyone's own needs.

However, I would suggest that in addition to "Property" - "Qualifier" -> "Value", there would also be "Scope" in front of them. Experience has taught that property will probably have a simple name, and when merging content from multiple places, it almost certainly will collide, and when it does, the values are not in any way mappable. Basically, the Scope could have a value like `Aalto`, `HELM` or `IDEMS` and would then tell us directly that whatever the property is, it should be interpreted by the rules of that organisation/group/project. And I do want scope as a separate item in the JSON so that queries do not need to extract it from some syntax mixed into the property name.

**Comment by aharjula** [comment]
Well, the user would have less work to do if there were a shared scope for a set of properties. All I ask is that there can be multiple such sets of properties with different scopes. Possibly, one would not have "Additional Metadata" visible at the start; one would add it as a set and, while adding, declare the scope for that set of metadata.

The UI might be less vertically heavy if each row had a scope, but it would also require those scope fields to be filled to be of any use. No one is going to remember to fill them and will simply assume that everyone understands if they simply fill the first rows one...

I would not want to be forced to convert or drop meta when a material with a different scope appears in my world; I would rather write a translated and locally expanded scoped metadata set alongside the old one.

**Comment by christianp** [comment]
Why not choose "all rights reserved" as the default licence, to avoid catching out people who haven't configured the setting?

**Comment by ProfDrMike** [comment]
This looks really good! A few comments:

“creator” should also be allowed to represent a non-natural person. This may be relevant for licensing purposes. The case is already covered if “firstName” and “lastName” are not mandatory fields. In that case, the non-natural person would appear under “institution”, while the actual authors would appear under “contributor”.

“license” could additionally include fields for attribution and a note on how the attribution should be provided.

Attached is a JSON example. In this example, taxonomy is implemented as “scope”. Alternatively, it could be integrated into the main section (metadata header) if taxonomy is intended to be emphasized and subject/topic should appear more prominently (as part of the taxonomy).

{
  "creator": {
    "firstName": "",
    "lastName": "",
    "institution": "DZdA",
    "year": "2026"
  },
  "contributor": [
    {
      "firstName": "Tom",
      "lastName": "Smith",
      "institution": "MIT",
      "year": "2026"
    }
  ],
  "language": [
    "en",
    "de"
  ],
  "isPartOf": "DZdA",
  "license": "CC BY-NC 4.0",
  "licenseInfos": {
    "attribution": "Deutsches Zentrum für digitale Aufgaben in der Hochschullehre (DZdA)",
    "attributionLink": "https://dzda.info",
    "licenseLink": "https://creativecommons.org/licenses/by-nc/4.0/"
  },
  "additional": {
    "taxonomy": {
      "taxonomyScheme": "DZdA-Math"
	  "content": {
        "topic": "Fundamentals",
        "subtopic": "Sets",
        "specificContents": [
          "Cardinality",
          "Set difference",
          "Venn diagram"
        ]
      },
      "difficulty": "Medium",
      "taskType": "Comprehension task"
    }
  }
}

**Comment by EJMFarrow** [comment]
Thanks, @ProfDrMike . Point taken about creator/contributor. I've removed the requirement for a lastName in the form.

For additional license info, users can use the addtional fields and then they can put in whatever they want. With input from @m-r-k the design has now been updated to output more natural JSON. Documented here https://github.com/maths/moodle-qtype_stack/blob/iss1171/doc/en/Authoring/Metadata.md

There's also a `freeform` property where power users can enter valid JSON in whatever layout they need e.g. deeper nesting. Other users can stick to the more restrictive `additional` property which is constrained by the form layout. 

```
{
    "creator": {
        "firstName": "Dave",
        "lastName": "Summers",
        "institution": "Edinburgh",
        "year": "2025"
    },
    "contributor": [
        {
            "firstName": "Bob",
            "lastName": "Smith",
            "institution": "Open University",
            "year": "2026"
        },
        {
            "firstName": "Tim",
            "lastName": "Jones",
            "institution": "",
            "year": "2026"
        }
    ],
    "language": [
        "en",
        "de"
    ],
    "isPartOf": "HELM",
    "license": "cc-4.0",
    "additional": {
        "Edinburgh": {
            "Course": {
                "Name": "Introductory Maths",
                "Week": "3",
                "Code": [
                    "AA",
                    "BB"
                ]
            },
            "Topic": [
                "Calculus",
                "Geometry",
                "Statistics"
            ],
            "Level": "Undergraduate"
        },
        "HELM": {
            "Workbook": "10",
            "Level": "Basic"
        }
    },
    "freeform": {}
}
```
